src/librustc_target/abi/call/riscv.rs

```diff
@@ -1,49 +1,341 @@
+
+#[derive(Copy, Clone)]
+enum FieldKind {
+    Float,
+    Integer,
+    Unknown,
+}
+
+#[derive(Copy, Clone)]
+enum FloatArgPass {
+    FloatPair,
+    Float,
+    MixedPair,
+}
+
+fn is_vector<'a, Ty>(arg: &ArgAbi<'a, Ty>) -> bool {
+    match arg.layout.abi {
+        Abi::Vector { .. } => true,
+        _ => false,
+    }
+}
+
+fn is_aggregate<'a, Ty>(arg: &ArgAbi<'a, Ty>) -> bool {
+    arg.layout.is_aggregate() || is_vector(arg)
+}
+
+fn should_use_fp_conv_helper<'a, Ty, C>(
+    cx: &C,
+    arg_layout: &TyLayout<'a, Ty>,
+    xlen: u64,
+    flen: u64,
+    field1_kind: &mut FieldKind,
+    field2_kind: &mut FieldKind,
+) -> bool
+where
+    Ty: TyLayoutMethods<'a, C> + Copy,
+    C: LayoutOf<Ty = Ty, TyLayout = TyLayout<'a, Ty>>,
+{
+    match arg_layout.abi {
+        Abi::Scalar(ref scalar) => match scalar.value {
+            abi::Int(..) | abi::Pointer => {
+                if arg_layout.size.bits() > xlen {
+                    return false;
+                }
+                match (*field1_kind, *field2_kind) {
+                    (FieldKind::Unknown, _) => {
+                        *field1_kind = FieldKind::Integer;
+                        true
+                    }
+                    (FieldKind::Float, FieldKind::Unknown) => {
+                        *field2_kind = FieldKind::Integer;
+                        true
+                    }
+                    _ => false,
+                }
+            }
+            abi::F32 | abi::F64 => {
+                if arg_layout.size.bits() > flen {
+                    return false;
+                }
+                match (*field1_kind, *field2_kind) {
+                    (FieldKind::Unknown, _) => {
+                        *field1_kind = FieldKind::Float;
+                        true
+                    }
+                    (_, FieldKind::Unknown) => {
+                        *field2_kind = FieldKind::Float;
+                        true
+                    }
+                    _ => false,
+                }
+            }
+        },
+        Abi::Vector { .. } | Abi::Uninhabited => false,
+        Abi::ScalarPair(..) | Abi::Aggregate { .. } => {
+            let is_union = match arg_layout.fields {
+                FieldPlacement::Union(_) => true,
+                FieldPlacement::Array { count, .. } => {
+                    for _ in 0..count {
+                        let elem_layout = arg_layout.field(cx, 0);
+                        if !should_use_fp_conv_helper(
+                            cx,
+                            &elem_layout,
+                            xlen,
+                            flen,
+                            field1_kind,
+                            field2_kind,
+                        ) {
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+                FieldPlacement::Arbitrary { .. } => false,
+            };
```

Comment:
So this is like the homogenous aggregate computation?
Can you make this function a method on a `struct` with all of the arguments as fields, except `arg_layout`? I feel like it would be a bit easier to read then.